module/VuFind/tests/unit-tests/src/VuFindTest/RateLimiter/Turnstile/TurnstileTest.php

```diff
@@ -0,0 +1,157 @@
+
+        $event->setRouteMatch(new RouteMatch(['controller' => 'SearchApi']));
+        $this->assertFalse($turnstile->isChallengeAllowed($event));
+
+        $event->setRouteMatch(new RouteMatch(['controller' => 'Cover']));
+        $this->assertFalse($turnstile->isChallengeAllowed($event));
+
+        // Other example routes
+        $event->setRouteMatch(new RouteMatch(['controller' => 'Search']));
+        $this->assertTrue($turnstile->isChallengeAllowed($event));
+
+        $event->setRouteMatch(new RouteMatch(['controller' => 'Record']));
+        $this->assertTrue($turnstile->isChallengeAllowed($event));
+    }
+
+    /**
+     * Test validating and returning positive and negative results,
+     * and test reporting that there was no prior result.
+     *
+     * @return void
+     */
+    public function testValidateAndCacheResults()
+    {
+        $policyId = 'policy1';
+        $clientIp = '111.222.123.213';
+        foreach ([true, false, null] as $result) {
+            $turnstile = $this->buildTurnstile($result);
+
+            if ($result != null) {
+                $turnstile->setHttpService($this->buildHttpService(
+                    ['success' => $result]
+                ));
+                $validationResult = $turnstile->validateToken('some_token', $policyId, $clientIp);
+                $this->assertEquals($result, $validationResult);
+            }
+
+            $priorResult = $turnstile->checkPriorTurnstileResult($policyId, $clientIp);
+            $this->assertEquals($result, $priorResult);
+        }
+    }
+
+    /**
+     * Build a Turnstile object.
+     *
+     * @param $cacheResult The result to report for a cache lookup.
+     *
+     * @return Turnstile
+     */
+    protected function buildTurnstile($cacheResult = null): Turnstile
+    {
+        $config = [
+            'Policies' => [
+                'policy1' => [
+                    'turnstileRateLimiterSettings' => [
+                        'limit' => '100',
+                    ],
+                ],
+            ],
+            'Turnstile' => [
+                'secretKey' => '12345,',
+            ],
+        ];
+
+        $cache = $this->getMockBuilder(\Laminas\Cache\Storage\StorageInterface::class)->getMock();
+        $cache->method('getItem')->willReturn($cacheResult);
+
+        $turnstile = new Turnstile($config, $cache);
+        return $turnstile;
+    }
+
+    /**
+     * Build a mock HttpService that returns a JSON
+     * representation of the given data.
+     *
+     * @param $returnData The data
+     *
+     * @return HttpService
+     */
+    protected function buildHttpService($returnData): HttpService
+    {
+        $httpService = $this->getMockBuilder(HttpService::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+        $response = $this->getMockBuilder(Response::class)
+            ->disableOriginalConstructor()
+            ->getMock();
```

Comment:
This can be simplified:

```suggestion
        $httpService = $this->createMock(HttpService::class);
        $response = $this->createMock(Response::class);
```